File: projects/PROJ-046-the-relationship-between-sleep-chronotyp/code/05_validate_report.py
import os
import re
import sys
from pathlib import Path
from typing import List, Tuple, Set

def get_project_root() -> Path:
    """Return the project root directory."""
    # Assuming the script is run from the project root or code directory
    # We look for a specific marker or traverse up
    current = Path(__file__).resolve()
    # Traverse up until we find a directory that looks like the root
    # For PROJ-046, we assume the root is the parent of 'code'
    if current.name == 'code':
        return current.parent
    parent = current.parent
    while parent != parent.parent:
        if (parent / 'tasks.md').exists():
            return parent
        parent = parent.parent
    return current.parent

def read_report_content(report_path: Path) -> str:
    """Read the content of the generated report (HTML or PDF text extraction)."""
    if not report_path.exists():
        raise FileNotFoundError(f"Report file not found: {report_path}")
    
    # For HTML reports, read raw text
    if report_path.suffix.lower() == '.html':
        return report_path.read_text(encoding='utf-8')
    
    # For PDF, we would need a library like PyPDF2, but for this validation
    # we assume the primary output is HTML as per T028 description.
    # If PDF is required, this logic would need expansion.
    raise ValueError(f"Unsupported report format: {report_path.suffix}. Expected .html")

def validate_report(report_content: str) -> Tuple[bool, List[str]]:
    """
    Parse report content and verify presence of all required sections.
    
    Required sections (SC-003):
    1. Descriptive Statistics
    2. ANCOVA Results
    3. Effect Sizes (Cohen's d)
    4. Power Analysis (G*Power)
    5. Sensitivity Analysis
    6. Cronbach's Alpha / Reliability Metrics
    7. Data Exclusion Summary
    
    Returns:
        Tuple[bool, List[str]]: (is_valid, list_of_missing_sections)
    """
    required_sections = [
        r"Descriptive\s+Statistics",
        r"ANCOVA\s+Results",
        r"Effect\s+Size",
        r"Power\s+Analysis",
        r"Sensitivity\s+Analysis",
        r"Cronbach",
        r"Exclusion"
    ]
    
    missing = []
    content_lower = report_content.lower()
    
    # Normalize whitespace for regex matching
    normalized_content = re.sub(r'\s+', ' ', content_lower)
    
    for pattern in required_sections:
        if not re.search(pattern, normalized_content, re.IGNORECASE):
            missing.append(pattern)
    
    return len(missing) == 0, missing

def main():
    project_root = get_project_root()
    report_path = project_root / "reports" / "chronotype_moral_analysis.html"
    
    if not report_path.exists():
        print(f"ERROR: Report file not found at {report_path}")
        print("Ensure code/04_report.Rmd has been rendered to reports/chronotype_moral_analysis.html")
        sys.exit(1)
    
    try:
        content = read_report_content(report_path)
        is_valid, missing_sections = validate_report(content)
        
        if is_valid:
            print("SUCCESS: Report validation passed. All required sections present.")
            sys.exit(0)
        else:
            print("FAILURE: Report validation failed.")
            print(f"Missing sections: {', '.join(missing_sections)}")
            sys.exit(1)
            
    except FileNotFoundError as e:
        print(f"ERROR: {e}")
        sys.exit(1)
    except Exception as e:
        print(f"ERROR: Unexpected error during validation: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: projects/PROJ-046-the-relationship-between-sleep-chronotyp/code/05_validate_report.R
#!/usr/bin/env Rscript
# code/05_validate_report.R
# Validates the generated R-Markdown report for required sections and content
#
# This script ensures that the final analysis report contains all required sections
# as specified in the project requirements (SC-003).

library(readr)
library(stringr)
library(dplyr)

# Constants
REQUIRED_SECTIONS <- c(
  "Descriptive Statistics",
  "Chronotype Classification Summary",
  "Reliability Metrics (Cronbach's Alpha)",
  "ANCOVA Results",
  "Effect Sizes (Cohen's d)",
  "Power Analysis",
  "Sensitivity Analysis",
  "Data Exclusion Summary",
  "Conclusions"
)

REQUIRED_COLUMNS_SENSITIVITY <- c(
  "subscale",
  "alpha_threshold",
  "p_value",
  "significant"
)

MIN_ALPHA_THRESHOLDS <- 3

# Function to read report content
read_report_content <- function(report_path) {
  if (!file.exists(report_path)) {
    stop(paste("Report file not found:", report_path))
  }

  # Read HTML content
  content <- read_lines(report_path, warn = FALSE)
  paste(content, collapse = "\n")
}

# Function to validate report structure
validate_report <- function(report_path) {
  cat("Validating report:", report_path, "\n")

  # Read report content
  content <- read_report_content(report_path)

  # Check for required sections
  missing_sections <- c()
  for (section in REQUIRED_SECTIONS) {
    if (!str_detect(content, str_c("\\b", section, "\\b", ignore_case = TRUE))) {
      missing_sections <- c(missing_sections, section)
    }
  }

  if (length(missing_sections) > 0) {
    cat("ERROR: Missing required sections:\n")
    for (section in missing_sections) {
      cat("  -", section, "\n")
    }
    return(FALSE)
  }

  cat("✓ All required sections present\n")

  # Validate sensitivity analysis table
  if (file.exists("data/derived/sensitivity_sweep.csv")) {
    sensitivity_data <- read_csv("data/derived/sensitivity_sweep.csv", show_col_types = FALSE)

    # Check required columns
    missing_cols <- setdiff(REQUIRED_COLUMNS_SENSITIVITY, names(sensitivity_data))
    if (length(missing_cols) > 0) {
      cat("ERROR: Sensitivity table missing columns:\n")
      for (col in missing_cols) {
        cat("  -", col, "\n")
      }
      return(FALSE)
    }

    # Check minimum alpha thresholds
    unique_thresholds <- unique(sensitivity_data$alpha_threshold)
    if (length(unique_thresholds) < MIN_ALPHA_THRESHOLDS) {
      cat("ERROR: Sensitivity table has", length(unique_thresholds),
          "alpha thresholds, but requires at least", MIN_ALPHA_THRESHOLDS, "\n")
      return(FALSE)
    }

    cat("✓ Sensitivity analysis table valid\n")
    cat("  - Columns:", paste(REQUIRED_COLUMNS_SENSITIVITY, collapse = ", "), "\n")
    cat("  - Alpha thresholds:", length(unique_thresholds), "\n")
  } else {
    cat("WARNING: Sensitivity sweep file not found (data/derived/sensitivity_sweep.csv)\n")
  }

  # Validate reliability metrics
  if (file.exists("data/derived/reliability_metrics.csv")) {
    reliability_data <- read_csv("data/derived/reliability_metrics.csv", show_col_types = FALSE)

    required_rel_cols <- c("subscale", "cronbach_alpha", "n_items")
    missing_rel_cols <- setdiff(required_rel_cols, names(reliability_data))

    if (length(missing_rel_cols) > 0) {
      cat("ERROR: Reliability metrics missing columns:\n")
      for (col in missing_rel_cols) {
        cat("  -", col, "\n")
      }
      return(FALSE)
    }

    # Check for valid alpha values
    invalid_alphas <- sum(reliability_data$cronbach_alpha < 0 | reliability_data$cronbach_alpha > 1)
    if (invalid_alphas > 0) {
      cat("ERROR: Invalid Cronbach's alpha values found (must be between 0 and 1)\n")
      return(FALSE)
    }

    cat("✓ Reliability metrics valid\n")
  } else {
    cat("WARNING: Reliability metrics file not found (data/derived/reliability_metrics.csv)\n")
  }

  # Validate ANCOVA results
  if (file.exists("data/derived/ancova_results.csv")) {
    ancova_data <- read_csv("data/derived/ancova_results.csv", show_col_types = FALSE)

    required_ancova_cols <- c("subscale", "chronotype_contrast", "p_value", "p_value_corrected")
    missing_ancova_cols <- setdiff(required_ancova_cols, names(ancova_data))

    if (length(missing_ancova_cols) > 0) {
      cat("ERROR: ANCOVA results missing columns:\n")
      for (col in missing_ancova_cols) {
        cat("  -", col, "\n")
      }
      return(FALSE)
    }

    cat("✓ ANCOVA results valid\n")
  } else {
    cat("WARNING: ANCOVA results file not found (data/derived/ancova_results.csv)\n")
  }

  # Validate effect sizes
  if (file.exists("data/derived/effect_sizes.csv")) {
    effect_data <- read_csv("data/derived/effect_sizes.csv", show_col_types = FALSE)

    required_effect_cols <- c("subscale", "contrast", "cohen_d", "ci_lower", "ci_upper")
    missing_effect_cols <- setdiff(required_effect_cols, names(effect_data))

    if (length(missing_effect_cols) > 0) {
      cat("ERROR: Effect sizes missing columns:\n")
      for (col in missing_effect_cols) {
        cat("  -", col, "\n")
      }
      return(FALSE)
    }

    cat("✓ Effect sizes valid\n")
  } else {
    cat("WARNING: Effect sizes file not found (data/derived/effect_sizes.csv)\n")
  }

  cat("\n✓ Report validation PASSED\n")
  return(TRUE)
}

# Main execution
main <- function() {
  args <- commandArgs(trailingOnly = TRUE)

  # Default report path
  report_path <- "reports/chronotype_moral_analysis.html"

  # Allow override via command line
  if (length(args) >= 1) {
    report_path <- args[1]
  }

  # Validate report
  is_valid <- validate_report(report_path)

  # Exit with appropriate code
  if (!is_valid) {
    cat("\n✗ Report validation FAILED\n")
    quit(status = 1)
  }

  cat("\n✓ All validations passed successfully\n")
  quit(status = 0)
}

# Run main if executed directly
if (!interactive()) {
  main()
}